### DISCUSSION TITLE
Which keys are pressed during the Alt+Tab Swapper action

### DISCUSSION THREAD

**Discussion by ziombo** [urob/zmk-config#87] [Q&A] [answered]
I believe I'm using a slightly older implementation than what you currently have, but the new one seems to be just a wrapper.

Mine
```
ZMK_BEHAVIOR(swapper, tri_state,
    bindings = <&kt LALT>, <&kp TAB>, <&kt LALT>;
    ignored-key-positions = <LT2>;
)
```

Current
```
ZMK_TRI_STATE(swapper,
    bindings = <&kt LALT>, <&kp TAB>, <&kt LALT>;
    ignored-key-positions = <LT2>;
)
```

I'm wondering what exactly happens in terms of keypresses when I trigger the above. I'm asking because I recently started using Parsec (remote desktop, but for gaming so the latency and quality are great) to connect to my work laptop and it has an 'Immersed' mode in which even the alt+tab action is passed into the remote machine. In other words when pressing alt+tab instead of switching from Parsec window to the next one on my local pc, I can switch between programs on my work laptop to which I'm connected. Using alt+tab directly works without issues.

However there seems to be an issue when using the swapper. The first time I press the swapper key it doesn't trigger alt+tab inside parsec and instead a normal Tab is received. The second time I do it it works as expected. So that's why I'm asking about the swagger's course of action and keypresses. Also holding down the swapper key will result in a Tab keypress and then switching windows.

Based on the config `<&kt LALT>, <&kp TAB>, <&kt LALT>` will toggle LALT, then press TAB and then release LALT? Would you have any ideas why would the first press be registered as just Tab?

**Comment by urob** [comment]
This is the key events it sends:
- press `&kp LALT` (when behavior starts) 
- press and release `&kp TAB` (when behavior starts & then again each time you press `&swapper`)
- release `&kp LALT`(when behavior ends)

**Comment by ziombo** [comment]
My first thought is that maybe the delay between first LALT and TAB is too small so it doesn't register it correctly. Is the delay between those actions configured somewhere and would it be possible to take control over that?

**Reply by ziombo** [reply]
I have now managed to replicate this with pressing LALT and TAB in quick succession, which behaves in the same way - just the Tab signal is processed. I believe this is more related to how ZMK treats the keypresses and sends them further, but it brings me back to the question - would it be possible to configure a delay between the swapper bindings?

**Reply by urob** [reply]
`tri-state` itself doesn't have a way to delay the tab. But you could try replacing the first or second binding with macro that adds a gap. 

Replacing the second binding with a macro that adds a gap before `TAB` should work for sure at the expanse of adding it also for subsequent TAB-presses. Replacing the first binding with a macro that adds a gap after the `LALT` wouldn't have that drawback but I'm not a hundred percent sure this would succeed in delaying triggering the second binding.

**Reply by ziombo** [reply]
Is this what you have in mind

```
ZMK_BEHAVIOR(delayed_tab, macro,
    wait-ms = <500>;
    tap-ms = <5>;
    bindings = <&kp TAB>;
)

// Alt+Tab swapper, requires PR #1366
ZMK_BEHAVIOR(swapper, tri_state,
    bindings = <&kt LALT>, <&delayed_tab>, <&kt LALT>;
    ignored-key-positions = <LT2>;
)
```

I see the delay in general, but it hasn't affected the issue, still on first swapper use the tab is received without the ALT.


By some trial an error I managed to find something that makes it work and seems to be working like the original swapper:
```
ZMK_BEHAVIOR(delayed_tab, macro,
    wait-ms = <10>;
    tap-ms = <1>;
    bindings = <&kp LALT>
             , <&kp TAB>;
)

// Alt+Tab swapper, requires PR #1366
ZMK_BEHAVIOR(swapper, tri_state,
    bindings = <&kt LALT>, <&delayed_tab>, <&kt LALT>;
    ignored-key-positions = <LT2>;
)
```
so basically an additional keypress of LALT and since it's already toggled I don't think it makes any difference. Unless you see an obvious issue with that I'd treat the issue as solved

**Reply by ziombo** [reply]
Unfortunately it seems that the LALT signals aren't ending properly , but I think that's the general direction for me to look into. Thanks for the help (also thanks for chiming in on my question on ZMK discord 😄 )

**Reply by urob** [reply, marked as answer]
Yea, that looks a bit weird. Effectively you are getting:
- Press `LALT` when the behavior is first invoked
- Press `LALT` again and then release it on every subsequent behavior tap
- Release `LALT` when the behavior is released.

I think the other approach to delay the tap might be more promising. The problem is that afaik `wait-ms` only operates _in between_ macro actions. Not sure if there is a more elegant way but you could try adding a dummy event at the beginning to get around that. Maybe something like this works?

```c
ZMK_BEHAVIOR(delayed_tab, macro,
    wait-ms = <500>;
    tap-ms = <5>;
    bindings = <&none &kp TAB>;
)

// Alt+Tab swapper, requires PR #1366
ZMK_BEHAVIOR(swapper, tri_state,
    bindings = <&kt LALT>, <&delayed_tab>, <&kt LALT>;
    ignored-key-positions = <LT2>;
)
```

**Reply by ziombo** [reply]
Sorry for the long delay. This appears to be working and the crucial part is the wait-ms setting, too short and it doesn't catch the combo. Thanks again!
